File: bindings/kotlin/scp-kt/src/test/kotlin/works/limn/scp/OutletStreamVectorsTest.kt
// SCP-OUT-039 — Outlet streaming vector smoke tests (Kotlin SDK).
//
// Loads the seven streaming conformance vectors at
// `tests/conformance/vectors/outlet_stream_vectors.json` and drives
// each through an `InvocationHandle` pump, asserting the vector's
// declared terminal-status surface reproduces under the SDK control
// plane.
//
// Per SCP-OUT-039 AC6: each vector runs in each SDK and produces the
// expected terminal status. Runtime-side replay (CreditTracker /
// CancelAckTracker / StreamEscrow) lives in
// `crates/scp-testing/tests/integration/outlet_stream_conformance.rs`;
// this smoke ensures the Kotlin SDK can ingest the same JSON vectors
// and reproduce the surface-level outcome.
//
// The cancellation, credit-exhaustion and sequence-gap vectors all
// terminate with a terminal Error chunk via the Flow surface — the
// wire-level distinction between "framework-emitted cancel-ack" and
// "receiver-emitted StreamGap" is a runtime concern.

package works.limn.scp

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import org.junit.jupiter.api.Test

@Suppress("StringLiteralDuplication", "TooManyFunctions")
class OutletStreamVectorsTest {
    @Serializable
    private data class VectorFile(
        val comment: String,
        @kotlinx.serialization.SerialName("spec_section")
        val specSection: String,
        val vectors: List<Vector>,
    )

    @Serializable
    private data class Vector(
        val name: String,
        val description: String,
        val open: OpenSpec,
        val chunks: List<ChunkEntry>,
        @kotlinx.serialization.SerialName("expected_end_status")
        val expectedEndStatus: String,
        @kotlinx.serialization.SerialName("expected_error_code")
        val expectedErrorCode: String? = null,
        @kotlinx.serialization.SerialName("expected_error_slug")
        val expectedErrorSlug: String? = null,
        @kotlinx.serialization.SerialName("expected_chunks_billed")
        val expectedChunksBilled: Int,
        @kotlinx.serialization.SerialName("expected_total_chunks")
        val expectedTotalChunks: Int,
        @kotlinx.serialization.SerialName("expected_cancel_ack_seq")
        val expectedCancelAckSeq: Long? = null,
        @kotlinx.serialization.SerialName("expected_first_gap_sequence")
        val expectedFirstGapSequence: Long? = null,
    )

    @Serializable
    private data class OpenSpec(
        @kotlinx.serialization.SerialName("outlet_id") val outletId: String,
        @kotlinx.serialization.SerialName("outlet_kind") val outletKind: String,
        // input is JSON-shaped, captured as JsonElement.
        val input: JsonElement,
        @kotlinx.serialization.SerialName("invoker_did") val invokerDid: String,
        @kotlinx.serialization.SerialName("operator_did") val operatorDid: String,
        @kotlinx.serialization.SerialName("context_id") val contextId: String,
        @kotlinx.serialization.SerialName("credit_window") val creditWindow: Int,
        @kotlinx.serialization.SerialName("estimated_chunk_count") val estimatedChunkCount: Int,
        @kotlinx.serialization.SerialName("cost_per_chunk") val costPerChunk: Long,
        @kotlinx.serialization.SerialName("available_balance") val availableBalance: Long,
        @kotlinx.serialization.SerialName("stream_credit_stall_secs") val streamCreditStallSecs: Int,
        @kotlinx.serialization.SerialName("stream_cancel_ack_secs") val streamCancelAckSecs: Int,
        @kotlinx.serialization.SerialName("timeout_ms") val timeoutMs: Int,
        @kotlinx.serialization.SerialName("chain_depth") val chainDepth: Int,
    )

    @Serializable
    private data class ChunkEntry(
        val sequence: Long,
        val type: String,
        val value: JsonElement? = null,
        val aggregate: JsonElement? = null,
        @kotlinx.serialization.SerialName("execution_time_ms")
        val executionTimeMs: Long? = null,
        val code: String? = null,
        val message: String? = null,
        val terminal: Boolean? = null,
        val pct: Int? = null,
        val note: String? = null,
        val slug: String? = null,
    )

    private val json = Json { ignoreUnknownKeys = true }

    private fun vectorPath(): Path {
        var dir = Paths.get("").toAbsolutePath()
        repeat(8) {
            val candidate = dir.resolve("tests/conformance/vectors/outlet_stream_vectors.json")
            if (Files.exists(candidate)) return candidate
            val parent = dir.parent ?: return@repeat
            dir = parent
        }
        error("outlet_stream_vectors.json not found from ${'$'}{Paths.get(\"\").toAbsolutePath()}")
    }

    @OptIn(ExperimentalSerializationApi::class)
    private fun loadVectors(): List<Vector> {
        val bytes = Files.readAllBytes(vectorPath())
        return json.decodeFromString(VectorFile.serializer(), String(bytes, Charsets.UTF_8)).vectors
    }

    private val requestId: ByteArray = ByteArray(16) { 0xa5.toByte() }

    private fun chunkFromEntry(entry: ChunkEntry): OutletStreamChunk =
        when (entry.type) {
            "data" -> OutletStreamChunk.Data(
                requestId = requestId,
                sequence = entry.sequence,
                valueJson = entry.value?.toString() ?: "null",
            )
            "end" -> OutletStreamChunk.End(
                requestId = requestId,
                sequence = entry.sequence,
                aggregateJson = entry.aggregate?.toString() ?: "null",
                executionTimeMs = entry.executionTimeMs ?: 0L,
            )
            "error" -> OutletStreamChunk.Error(
                requestId = requestId,
                sequence = entry.sequence,
                code = entry.code ?: "SCP-TOOL-6200",
                message = entry.message ?: "",
                terminal = entry.terminal ?: false,
            )
            "progress" -> OutletStreamChunk.Progress(
                requestId = requestId,
                sequence = entry.sequence,
                pct = (entry.pct ?: 0).toUShort(),
                note = entry.note,
            )
            else -> error("unknown chunk type: ${'$'}{entry.type}")
        }

    private fun makeHandle(vector: Vector): InvocationHandle {
        val sdkChunks = vector.chunks.map(::chunkFromEntry).toMutableList()
        if (vector.name == "sequence_gap") {
            // Synthesize the receiver's terminal StreamGap envelope so
            // the Flow can drain. The runtime's actual StreamGap path
            // is exercised in the Rust replay.
            sdkChunks += OutletStreamChunk.Error(
                requestId = requestId,
                sequence = vector.chunks.last().sequence + 1,
                code = vector.expectedErrorCode ?: "SCP-TOOL-6131",
                message = vector.expectedErrorSlug ?: "execution.stream-gap",
                terminal = true,
            )
        }

        val endChunk = sdkChunks.filterIsInstance<OutletStreamChunk.End>().firstOrNull()

        return InvocationHandle(
            aggregateFn = {
                endChunk?.let {
                    Aggregate(valueJson = it.aggregateJson, executionTimeMs = it.executionTimeMs)
                } ?: Aggregate(valueJson = "null")
            },
            flowFn = {
                flow {
                    for (chunk in sdkChunks) {
                        emit(chunk)
                    }
                }
            },
            requestIdHex = "a5".repeat(16),
            aggregateSchemaJson = null,
        )
    }

    private val requiredNames = setOf(
        "non_streaming",
        "multi_chunk",
        "cancellation",
        "error_terminal",
        "error_recoverable",
        "sequence_gap",
        "credit_exhaustion",
    )

    @Test
    fun `vector file carries exactly seven required vectors (AC1)`() {
        val names = loadVectors().map { it.name }.toSet()
        assertEquals(requiredNames, names)
    }

    @Test
    fun `each vector reproduces expected terminal status (AC6)`() = runTest {
        for (vector in loadVectors()) {
            val handle = makeHandle(vector)
            val observed: List<OutletStreamChunk> = handle.asFlow().toList()

            val expectedTotal = vector.expectedTotalChunks
            if (vector.name == "sequence_gap") {
                assertEquals(
                    expectedTotal + 1, observed.size,
                    "vector ${'$'}{vector.name}: observed = manifest + synthesized terminal",
                )
            } else {
                assertEquals(
                    expectedTotal, observed.size,
                    "vector ${'$'}{vector.name}: chunk count mismatch",
                )
            }

            val terminal = observed.last()
            when (vector.expectedEndStatus) {
                "Ok" -> {
                    assertTrue(
                        terminal is OutletStreamChunk.End,
                        "vector ${'$'}{vector.name}: terminal must be End",
                    )
                }
                "Error" -> {
                    assertTrue(
                        terminal is OutletStreamChunk.Error && terminal.terminal,
                        "vector ${'$'}{vector.name}: terminal must be terminal=true Error",
                    )
                    assertEquals(
                        vector.expectedErrorCode, terminal.code,
                        "vector ${'$'}{vector.name}: terminal Error code mismatch",
                    )
                }
                "Cancelled" -> {
                    assertTrue(
                        terminal is OutletStreamChunk.Error && terminal.terminal,
                        "vector ${'$'}{vector.name}: cancel-ack must surface as terminal Error",
                    )
                }
                else -> error(
                    "vector ${'$'}{vector.name}: unknown expected_end_status " +
                        "${'$'}{vector.expectedEndStatus}",
                )
            }
        }
    }

    @Test
    fun `every vector carries an input field on the open block`() {
        for (v in loadVectors()) {
            assertNotNull(v.open.input)
        }
    }

    @Test
    fun `vector file is well-formed JSON the SDK ingests as-is`() {
        val outer = json.decodeFromString(
            VectorFile.serializer(),
            Files.readAllBytes(vectorPath()).toString(Charsets.UTF_8),
        )
        assertEquals(7, outer.vectors.size)
        assertTrue(outer.specSection.isNotEmpty())
        for (v in outer.vectors) {
            assertTrue(v.chunks.isNotEmpty(), "vector ${'$'}{v.name}: chunks must be non-empty")
        }
    }
}

File: bindings/typescript/tests/outlet-stream-vectors.test.ts
/**
 * SCP-OUT-039 — Outlet streaming vector smoke tests (TypeScript SDK).
 *
 * Loads the seven streaming conformance vectors at
 * `tests/conformance/vectors/outlet_stream_vectors.json` and drives
 * each through an `InvocationHandle` pump, asserting the vector's
 * declared terminal-status surface reproduces under the SDK control
 * plane.
 *
 * Per SCP-OUT-039 AC6: each vector runs in each SDK and produces the
 * expected terminal status. Runtime-side replay (CreditTracker /
 * CancelAckTracker / StreamEscrow) lives in
 * `crates/scp-testing/tests/integration/outlet_stream_conformance.rs`;
 * this smoke ensures the TS SDK can ingest the same JSON vectors and
 * reproduce the surface-level outcome.
 *
 * The cancellation, credit-exhaustion and sequence-gap vectors all
 * terminate with a terminal Error chunk via the SDK iterator surface —
 * the wire-level distinction between "framework-emitted cancel-ack"
 * and "receiver-emitted StreamGap" is a runtime concern.
 */

import { describe, expect, test } from "bun:test";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

import { InvocationHandle, type OutletStreamChunk } from "../src/outlets";

// ---------------------------------------------------------------------------
// Vector loading
// ---------------------------------------------------------------------------

interface ChunkEntry {
  sequence: number;
  type: "data" | "end" | "error" | "progress";
  value?: unknown;
  aggregate?: unknown;
  execution_time_ms?: number;
  code?: string;
  message?: string;
  terminal?: boolean;
  pct?: number;
  note?: string | null;
  slug?: string;
}

interface VectorFile {
  comment: string;
  spec_section: string;
  vectors: Vector[];
}

interface Vector {
  name: string;
  description: string;
  open: {
    outlet_id: string;
    outlet_kind: string;
    input: unknown;
    invoker_did: string;
    operator_did: string;
    context_id: string;
    credit_window: number;
    estimated_chunk_count: number;
    cost_per_chunk: number;
    available_balance: number;
    stream_credit_stall_secs: number;
    stream_cancel_ack_secs: number;
    timeout_ms: number;
    chain_depth: number;
  };
  chunks: ChunkEntry[];
  credits: unknown[];
  cancel?: { after_sequence: number; expected_cancel_ack_seq: number };
  trigger?: string;
  expected_end_status: "Ok" | "Error" | "Cancelled";
  expected_error_code?: string | null;
  expected_error_slug?: string;
  expected_chunks_billed: number;
  expected_total_chunks: number;
  expected_cancel_ack_seq?: number | null;
  expected_first_gap_sequence?: number;
}

function vectorPath(): string {
  // Walk up from the test file to find the repo-root vectors file.
  let dir = resolve(import.meta.dir);
  for (let i = 0; i < 10; i++) {
    const candidate = resolve(dir, "tests/conformance/vectors/outlet_stream_vectors.json");
    try {
      readFileSync(candidate);
      return candidate;
    } catch {
      // not here — keep walking
    }
    const parent = resolve(dir, "..");
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error(`outlet_stream_vectors.json not found from ${import.meta.dir}`);
}

function loadVectors(): Vector[] {
  const raw = JSON.parse(readFileSync(vectorPath(), "utf8")) as VectorFile;
  expect(raw.vectors.length).toBe(7);
  return raw.vectors;
}

const REQUIRED_NAMES = [
  "non_streaming",
  "multi_chunk",
  "cancellation",
  "error_terminal",
  "error_recoverable",
  "sequence_gap",
  "credit_exhaustion",
] as const;

// ---------------------------------------------------------------------------
// Driver — feeds the InvocationHandle pump with the vector's chunks.
// ---------------------------------------------------------------------------

// Pump-side handler shape the InvocationHandle constructor accepts.
type PumpSink = Parameters<ConstructorParameters<typeof InvocationHandle>[0]>[0];

// Result of emitting one entry — `done` means the pump should stop.
type EmissionOutcome = "continue" | "done";

function emitEntry(sink: PumpSink, entry: ChunkEntry, requestId: Uint8Array): EmissionOutcome {
  if (entry.type === "data") {
    sink.chunk({
      requestId,
      sequence: entry.sequence,
      payloadType: "data",
      value: entry.value,
    } as OutletStreamChunk);
    return "continue";
  }
  if (entry.type === "end") {
    sink.end({
      value: entry.aggregate ?? null,
      ...(entry.execution_time_ms !== undefined && {
        executionTimeMs: entry.execution_time_ms,
      }),
    });
    return "done";
  }
  if (entry.type === "error") {
    sink.chunk({
      requestId,
      sequence: entry.sequence,
      payloadType: "error",
      code: entry.code ?? "SCP-TOOL-6200",
      message: entry.message ?? "",
      terminal: entry.terminal === true,
    } as OutletStreamChunk);
    if (entry.terminal === true) {
      sink.error(new Error(entry.message ?? "stream errored"));
      return "done";
    }
    return "continue";
  }
  if (entry.type === "progress") {
    sink.chunk({
      requestId,
      sequence: entry.sequence,
      payloadType: "progress",
      pct: entry.pct ?? 0,
      ...(entry.note != null && { note: entry.note }),
    } as OutletStreamChunk);
    return "continue";
  }
  return "continue";
}

function emitSequenceGapTerminal(sink: PumpSink, vector: Vector, requestId: Uint8Array): void {
  // sequence_gap vector intentionally omits a terminal — the
  // receiver's StreamGap cancel terminates from outside the
  // executor's emission. Synthesize the receiver's terminal Error so
  // the InvocationHandle iterator can drain.
  if (vector.name !== "sequence_gap") return;
  const lastSeq = vector.chunks[vector.chunks.length - 1]?.sequence ?? 0;
  sink.chunk({
    requestId,
    sequence: lastSeq + 1,
    payloadType: "error",
    code: vector.expected_error_code ?? "SCP-TOOL-6131",
    message: vector.expected_error_slug ?? "execution.stream-gap",
    terminal: true,
  } as OutletStreamChunk);
  sink.error(new Error("synthesized StreamGap"));
}

function buildHandle(vector: Vector): InvocationHandle {
  const requestId = new Uint8Array(16).fill(0xa5);
  return new InvocationHandle((sink) => {
    void Promise.resolve().then(() => {
      for (const entry of vector.chunks) {
        if (emitEntry(sink, entry, requestId) === "done") return;
      }
      emitSequenceGapTerminal(sink, vector, requestId);
    });
  });
}

async function drainHandle(vector: Vector): Promise<OutletStreamChunk[]> {
  const handle = buildHandle(vector);
  const observed: OutletStreamChunk[] = [];
  try {
    for await (const chunk of handle) {
      observed.push(chunk);
    }
  } catch {
    // The InvocationHandle pump path raises on terminal Error or
    // sequence-gap synthetic — that's the surface the SDK contract
    // exposes when the stream terminates abnormally. The chunks
    // already enqueued are visible; we proceed to assertions on them.
  }
  return observed;
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe("SCP-OUT-039 outlet stream vectors (TS smoke)", () => {
  test("vector file carries exactly the seven required vectors (AC1)", () => {
    const vectors = loadVectors();
    const names = new Set(vectors.map((v) => v.name));
    for (const required of REQUIRED_NAMES) {
      expect(names.has(required)).toBe(true);
    }
    expect(names.size).toBe(REQUIRED_NAMES.length);
  });

  test.each([
    ...REQUIRED_NAMES,
  ])("vector %s reproduces expected terminal status (AC6)", async (name: string) => {
    const v = loadVectors().find((x) => x.name === name);
    expect(v).toBeDefined();
    const vector = v as Vector;
    const observed = await drainHandle(vector);

    const expected_total = vector.expected_total_chunks;
    if (name === "sequence_gap") {
      // Smoke synthesizes one terminal Error on top of the manifest.
      expect(observed.length).toBe(expected_total + 1);
    } else {
      expect(observed.length).toBe(expected_total);
    }

    const terminal = observed[observed.length - 1];
    if (!terminal) {
      throw new Error(`vector ${name}: no chunks observed`);
    }
    switch (vector.expected_end_status) {
      case "Ok": {
        expect(terminal.payloadType).toBe("end");
        const endEntry = vector.chunks.find((c) => c.type === "end");
        expect(terminal.aggregate).toEqual(endEntry?.aggregate);
        break;
      }
      case "Error": {
        expect(terminal.payloadType).toBe("error");
        expect(terminal.terminal).toBe(true);
        expect(terminal.code).toBe(vector.expected_error_code ?? "");
        break;
      }
      case "Cancelled": {
        // Cancel-ack envelope per §5.4.5 — surfaces as a terminal
        // Error chunk to the SDK iterator. The runtime layer
        // distinguishes Cancelled vs Error on the
        // OutletInvokedEvent.
        expect(terminal.payloadType).toBe("error");
        expect(terminal.terminal).toBe(true);
        break;
      }
      default:
        throw new Error(`unknown expected_end_status ${vector.expected_end_status}`);
    }
  });

  test("every vector carries an input field on the open block", () => {
    for (const v of loadVectors()) {
      expect(v.open.input).toBeDefined();
    }
  });
});
